syntax = "proto3";
import "google/protobuf/descriptor.proto";  // нужно для работы индикаторов
package trainingV2;
option go_package = "../training";

// Методы которые должен имплементировать сервис провайдер
service Model {

  // free
  // можно прокинуть адрес создателя модели 
  rpc create_model(NewModel) returns (ModelID) {}

  // free
  rpc validate_model_price(ValidateRequest) returns (PriceInBaseUnit) {}

  // paid
  rpc upload_and_validate(stream DataToUpload) returns (StatusResponse) {}

  // paid
  rpc validate_model(ValidateRequest) returns (StatusResponse) {}

  // free, одна подпись для методов train_model_price & train_model
  rpc train_model_price(ModelID) returns (PriceInBaseUnit) {}

  // paid
  rpc train_model(ModelID) returns (StatusResponse) {}

  // free
  rpc delete_model(ModelID) returns (StatusResponse) {
    // После удаления модели status становится DELETED в etcd
  }

  // free
  rpc get_model_status(ModelID) returns (StatusResponse) {}
}

message ModelResponse {
  string model_id = 1;
  Status status = 2;
  string updated_date = 3;
  string name = 4;
  string description = 5;
  string grpc_method_name = 6;
  string grpc_service_name = 7;

  // List of all the addresses that will have access to this model
  repeated string address_list = 8;

  // доступ к модели даем только на использование и просмотр
  bool is_public = 9;

  string training_data_link = 10;
}

// Используется как input на запросы: new_model
// Использовать ли эти поля - выбор за сервис провайдером, главное вернуть model_id
message NewModel {
  string name = 1;
  string description = 2;
  string grpc_method_name = 3;
  string grpc_service_name = 4;

  // List of all the addresses that will have access to this model
  repeated string address_list = 5;

  // set this to true if you want your model to be used by other AI consumers
  bool is_public = 6;

  // Эти параметры будет прокидывать демон
  string organization_id = 7;
  string service_id = 8;
  string group_id = 9;
}

// Эту структуру нужно использовать сервис провайдеру
message ModelID {
  string model_id = 1;
}

// Эту структуру нужно использовать сервис провайдеру
// Используется в методе train_model_price для получения цены трейнинга/валидации
message PriceInBaseUnit {
  uint64 price = 1; // cogs, weis, afet, aasi etc
}

enum Status {
  CREATED = 0;
  VALIDATING = 1;
  VALIDATED = 2;
  TRAINING = 3;
  READY_TO_USE = 4; // after training completed
  ERRORED = 5;
  DELETED = 6;
}

message StatusResponse {
  Status status = 1;
}

message DataToUpload {
  string model_id = 2;
  bytes data = 3;
}

message ValidateRequest {
  string model_id = 2;
  string training_data_link = 3;
}

extend google.protobuf.MethodOptions {
  string defaultModelID = 50001;
  uint64 datasetMaxSizeMB = 50002;
  repeated string filesTypes = 50003; // [png, jpg, mp4]
  string datasetDescription = 50004; // свободное описание доп требований
  uint64 maxModelsPerUser = 50006;
}
